### DISCUSSION TITLE
Namespace qualify custom function

### DISCUSSION THREAD

**Discussion by kennyjwilli** [josdejong/mathjs#3203] [Q&A]
I'm curious if it is possible to namespace qualify a custom function I have exposed so it's not top-level. For example, I'd like to expose the `calculateMassWeight` function under the "namespace" `foo` so it is callable via `foo.calculateMassWeight`. 

```javascript
const mathjs = require('mathjs');

function calculateMassWeight(mass, gravity) {
  return mass * gravity;
}

const math = mathjs.create(mathjs.all);

math.import(
  {
    foo: {
      massWeight: calculateMassWeight,
    },
  },
  { override: true },
);

math.evaluate('foo.massWeight(10, 9.81)')
```

Calling the above `math.evaluate` results in the following error. 

```
Uncaught Error: Undefined symbol foo
    at Function.onUndefinedSymbol (/Users/kenny/work/kwllc/deeporigin/platform2/node_modules/.pnpm/mathjs@12.4.2/node_modules/mathjs/lib/cjs/expression/node/SymbolNode.js:221:15)
    at /Users/kenny/work/kwllc/deeporigin/platform2/node_modules/.pnpm/mathjs@12.4.2/node_modules/mathjs/lib/cjs/expression/node/SymbolNode.js:95:99
    at evalFunctionNode (/Users/kenny/work/kwllc/deeporigin/platform2/node_modules/.pnpm/mathjs@12.4.2/node_modules/mathjs/lib/cjs/expression/node/FunctionNode.js:263:26)
    at Object.evaluate (/Users/kenny/work/kwllc/deeporigin/platform2/node_modules/.pnpm/mathjs@12.4.2/node_modules/mathjs/lib/cjs/expression/node/Node.js:77:18)
    at Object.string (/Users/kenny/work/kwllc/deeporigin/platform2/node_modules/.pnpm/mathjs@12.4.2/node_modules/mathjs/lib/cjs/expression/function/evaluate.js:50:36)
    at Object.evaluate (/Users/kenny/work/kwllc/deeporigin/platform2/node_modules/.pnpm/typed-function@4.1.1/node_modules/typed-function/lib/umd/typed-function.js:1626:22)
```

It'd be great if, instead, I get the output of actually calling the `calculateMassWeight`.

**Comment by dvd101x** [comment]
Hi, in the meantime a better response is available, I found something odd.

Your import seems to work for `math.evaluate('massWeight(10, 9.81)')`

The docs states that one can import values and functions, so in theory it should be able to import your value that contains an object with a function.

It seems that if the value is an object, then it threats the object as the main object that contains values and functions.

```js
math.import({
    foo: {
        bar: 10
    }
})

math.foo     // currently undefined but should be {bar: 10}
math.foo.bar // currently throws an error but should be 10
math.bar     // currently 10 but should be undefined.
```

Maybe I'm misinterpreting the docs, but they state that you can import regular values, I've tried importing numbers, strings, units, arrays and objects but only numbers and strings works (I don't exactly know what regular values means)
 
I don't think it's possible to do it in that exact way with an import but you could use a scope or a parser for that.

```js
const scope = { 'foo': { massWeight: calculateMassWeight } }

math.evaluate('foo.massWeight(10, 9.81)', scope)

function calculateMassWeight(mass, gravity) {
    return mass * gravity;
}
```

```js
const parser = math.parser()

parser.set('foo', { massWeight: calculateMassWeight })

parser.evaluate('foo.massWeight(10, 9.81)')

function calculateMassWeight(mass, gravity) {
    return mass * gravity;
}
```

**Reply by josdejong** [reply]
Just using a `scope` like this is indeed the best option available I think 👍
